jfix-stdlib-ratelimiter/src/main/java/ru/fix/stdlib/ratelimiter/RateLimitedDispatcher.java

```diff
@@ -0,0 +1,211 @@
+        if (state != State.RUNNING) {
+            RejectedExecutionException ex = new RejectedExecutionException(
+                "RateLimiterDispatcher [" + name + "] is in [" + state + "] state"
+            );
+            result.completeExceptionally(ex);
+            return result;
+        }
+
+        @SuppressWarnings("resource")
+        ProfiledCall queueWaitTime = profiler.start("queue_wait");
+        queue.add(new Task(result, supplier, queueWaitTime));
+
+        return result;
+    }
+
+    public void updateRate(int rate) {
+        rateLimiter.updateRate(rate);
+    }
+
+    public void close(long timeoutMs) throws Exception {
+        boolean stateUpdated = state.compareAndSet(State.RUNNING, State.SHUTTING_DOWN);
+        if (!stateUpdated) {
+            logger.warn("Close called on RateLimitedDispatcher [{}] with state [{}]", name, state.get());
+            return;
+        }
+
+        thread.interrupt();
+        if (timeoutMs > 0) {
+            thread.join(timeoutMs);
+        }
+
+        stateUpdated = state.compareAndSet(State.SHUTTING_DOWN, State.TERMINATE);
+        if (!stateUpdated) {
+            logger.error(
+                "Can't set [TERMINATE] state to RateLimitedDispatcher [{}] in [{}] state",
+                name, state.get()
+            );
+            return;
+        }
+        thread.join();
+
+        rateLimiter.close();
+        profiler.detachIndicator(QUEUE_SIZE_INDICATOR);
+    }
+
+    @SuppressWarnings({"unchecked"})
+    private final class TaskProcessor implements Runnable {
+
+        @Override
+        public void run() {
+            while (state.get() == State.RUNNING ||
+                    state.get() == State.SHUTTING_DOWN && !queue.isEmpty()) {
+                try {
+                    processingCycle();
+                }
+                catch (Exception e) {
+                    logger.error(e.getMessage(), e);
+                }
+            }
+
+            queue.forEach(task -> {
+                task.getFuture().completeExceptionally(new RejectedExecutionException(
+                    "RateLimitedDispatcher [" + name + "] is in [TERMINATE] state"
+                ));
+                task.getQueueWaitTime().close();
+            });
+        }
+
+        private void processingCycle() {
+            Task task;
+            try {
+                task = queue.take();
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                return;
+            }
+
+            task.getQueueWaitTime().stop();
+            CompletableFuture future = task.getFuture();
+
+            try {
+                try (ProfiledCall limitAcquireTime = profiler.start("acquire_limit")) {
+
+                    boolean acquired = false;
+                    while (state.get() != State.TERMINATE && !acquired) {
+                        acquired = rateLimiter.tryAcquire(1, ChronoUnit.SECONDS);
```

Comment:
We can use `tryAcquire()` without timeout.  There is Thread.interrupt() invocation inside close method that will resume work in case of shutdown.   

Or at least maybe it would be better to increase timeout to reduce contention. E.g.. up to 10 seconds.